build/charts/antrea/templates/webhooks/validating/crdvalidator.yaml

```diff
@@ -60,7 +60,7 @@ webhooks:
     rules:
       - operations: ["CREATE", "UPDATE"]
         apiGroups: ["crd.antrea.io"]
-        apiVersions: ["v1alpha3", "v1alpha2"]
+        apiVersions: ["v1beta1", "v1alpha3"]
```

Comment:
My previous explanation of the API version used by the validation webhook may not be accurate. https://github.com/antrea-io/antrea/pull/5181#discussion_r1265861541 

Objects whose versions are explicitly listed here will be directly sent to the validation webhook with their original API versions intact. On the other hand, for objects with versions not listed here, they will first pass through the conversion webhook before being sent to the validation webhook.

For example, we list v1b1 and v1a3 here. If we create any v1b1 or v1a3 ClusterGroups, the validation webhook will receive these objects in their original API versions. However, if we create a v1a2 ClusterGroup, it will be converted to a v1a3 ClusterGroup by the conversion webhook. Subsequently, the validation webhook will receive the converted v1a3 ClusterGroup for validation.

I have a question about this: in general, is it ok to perform any conversions before validation?

- If it’s ok, then I’m thinking we could only list the latest version here, considering only the latest version will be stored in etcd and consumed by Antrea.
- If it’s not a good practice, then in the validation webhook, does unmarshaling different versions of objects into a specific version struct relying on the similarity still make sense?
Glad to hear your ideas. @tnqn @antoninbas

Reply:
> Objects whose versions are explicitly listed here will be directly sent to the validation webhook with their original API versions intact. On the other hand, for objects with versions not listed here, they will first pass through the conversion webhook before being sent to the validation webhook.

Do you any reference (documentation or code) that shows that this is true? Intuitively, I would have thought that validation would always happen before conversion, and that if the version was not listed here, it would not be validated at all.

Reply:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/admission/v1/types.go#L55

If we want to achieve "if the version was not listed here, it would not be validated at all.", then based on my understanding, we should change our `matchPolicy` to `Exact`. But according to the kubernetes doc, they seems don't recommend do this. 
https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#intercepting-all-versions-of-an-object

Reply:
Considering that, and after reading https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#intercepting-all-versions-of-an-object, it seems to me that your idea is a good one:

> If it’s ok, then I’m thinking we could only list the latest version here, considering only the latest version will be stored in etcd and consumed by Antrea.

I don't see a drawback in our case. In any case, v1alpha3 and v1beta1 are identical, so by definition we do not lose anything by converting first and then performing validation.

I guess doing conversion first, and then validation, could have an impact in some cases and could lead to user-facing error messages which are not as useful / correct. I don't believe it is the case here (or maybe it is for v1alpha2, but not in a way that matters).

I'll let @tnqn comment.

Reply:
The approach looks good to me.
I have a question though:
> However, if we create a v1a2 ClusterGroup, it will be converted to a v1a3 ClusterGroup by the conversion webhook. Subsequently, the validation webhook will receive the converted v1a3 ClusterGroup for validation.

How does kube-apiserver know which version it should convert the object to? Should it be v1b1 given it's the stored version?

Reply:
Sounds good. I pushed a commit changing the version in validation for ClusterGroup and Group only to include v1b1.
As for your question, I haven't found how apiserver chooses the version to convert before validation. I only found the definition of converting crd before storing to the etcd. https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L816

Reply:
Did you see it receives v1a3 version when you send a v1a2 version and the storage version is v1b1?

Reply:
Yes, if I put it like this `apiVersions: ["v1beta1", "v1alpha3"]`, then when I create a v1a2 cg, my dev log looks like below:
```
I0721 05:40:47.976536       1 convert.go:33] Converting CRD for ClusterGroup test-cg from crd.antrea.io/v1alpha2 to crd.antrea.io/v1alpha3
I0721 05:40:48.006805       1 validate.go:157] AdmissionReview received in NetworkPolicyValidator
I0721 05:40:48.006853       1 validate.go:158]     ar.Request.Kind: crd.antrea.io/v1alpha3, Kind=ClusterGroup
I0721 05:40:48.006873       1 validate.go:159]     ar.Request.RequestKind: crd.antrea.io/v1alpha2, Kind=ClusterGroup
I0721 05:40:48.010519       1 convert.go:33] Converting CRD for ClusterGroup test-cg from crd.antrea.io/v1alpha2 to crd.antrea.io/v1beta1
```
And for the storage:
```
╰─$ ku get cg -A -oyaml
apiVersion: v1
items:
- apiVersion: crd.antrea.io/v1beta1
  kind: ClusterGroup
  metadata:
  ...
```

Reply:
OK, I checked its implementation. It will use the first equivalent of the versions that match the rule. Since v1a3 is declared before v1b1 in the CRD, it will be used.  
https://github.com/kubernetes/kubernetes/blob/f5130e4387e068498e5d19b26cf6f7179482a9bb/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/webhook.go#L188-L196